[WIP] Clean-up source tree (OpenCL)

claudioandre opened the pull request:
### Summary

The source tree is "disorganized" with hundreds of "source files in a single
directory". For a starter, this is moving all OpenCL files (and nothing
else) to jumbo/opencl and jumbo/opencl/include.

### Other Information

* Passes CI;
* configure regenerated using autools;

### Other Information
- [ ] review
- [x] warning free
- [ ] what will happen to with Makefile legacy?

[edited]
* `squash` is mandatory.

solardiz commented:
Thank you for working on this, Claudio!

Should we perhaps postpone invasive changes like this until after the next numbered jumbo release (and make that release soon)?

claudioandre commented:
Well, it is not going to be ready soon. And, since I'm only moving OpenCL files from one place to another, I don't have a strong opinion about this. 

Anyway, I don't merge code I authored. So, it is up to someone else to merge (or not).

claudioandre commented on src/Makefile.in:
```diff
@@ -218,6 +220,7 @@ opencl_%.h:	| ../run/kernels
 
 ../run/kernels:
 	$(MKDIR) ../run/kernels
+	$(CP) jumbo/opencl/include/opencl_*.h ../run/kernels  #TODO: hack to make it work
```
Why is it needed? The copy above works for `.cl` files but not for `.h` files. Failed to see why!

claudioandre commented on src/configure.ac:
```diff
@@ -977,9 +985,11 @@ fi
   echo >fmt_registers.h
 fi
+cp arch.h jumbo/opencl/
+cp autoconfig.h jumbo/opencl/
```
Undesirable!

claudioandre commented on src/jumbo/opencl/Makefile.legacy:
```diff
@@ -0,0 +1,38 @@
+RM = rm -f
```
What should I do with this file?

magnumripper commented:
@solardiz wrote:
> Should we perhaps postpone invasive changes like this until after the next numbered jumbo release (and make that release soon)?

Personally I think not. I think we should merge now - there'll be plenty of time to iron out any last fixes while we test/fix the lots and lots of other issues.

claudioandre commented:
Rebased. To simplify review, I split the changes intro 3 commits:
. 1st: git mv;
. 3rd: autoreconf;
. 2nd is where the real stuff is.

My CI tested it on:
- OpenCL (clang + gcc)
- OpenCL "real cracking"
- no OpenCL (clang + gcc)
- with `--disable-native` + `--disable-openmp`
- specifying `CPPFLAGS`

I did:
- OpenCL and `--disable-opencl`
****
Squash them at will.

frank-dittrich commented:
> Squash them at will.

I think they must be squashed into 1 commit. Otherwise we end up with a commit which causes build errors (= trouble with `git bisect`).

For the review, the separation into 3 commits is OK.

magnumripper commented:
I was thinking the same. It's good for review, but we squash them when merging.

magnumripper commented:
Oh, the build warning is because of a bug. We check for HAVE_LIBZ without sourcing arch.h (which sources autoconfig.h) first. I just never noticed the format wasn't built.

magnumripper commented:
> what will happen to with Makefile legacy?

We need to fix it too before merging - or we could decide to ditch it entirely.

claudioandre commented:
>>what will happen to Makefile legacy?

>We need to fix it too (ideally before merging) - or we could decide to ditch it entirely.

My vote is to remove all OpenCL stuff from it. 

Well, unless there is someone eager to maintain it? I'm not; to me it's just duplicate work: autotools is mainstream.

frank-dittrich commented:
Let's wait at least a week, I might find time for this next week.

kholia commented:
I have to admit that my recent testing on Debian PPC64 VM was made possible by `Makefile.legacy`. It does have value.

magnumripper commented:
> Oh, the build warning is because of a bug in opencl_electrum_modern_fmt_plug.c. We check for HAVE_LIBZ without sourcing arch.h (which sources autoconfig.h) first. I just never noticed the format wasn't built.

@claudioandre I see now that bug was actually introduced by this PR (I though it was in bleeding). In f8debd25, you drop the wrong dupe.

magnumripper commented:
> @claudioandre I see now that bug was actually introduced by this PR (I though it was in bleeding). In f8debd2, you drop the wrong dupe.

I was wrong - it's not in this PR but that commit was made by you. IMO we can fix it in this PR (so you don't have to rebase again)

frank-dittrich commented:
I think that the sequence of
```
 #if HAVE_LIBZ
#ifdef HAVE_OPENCL
```
should be reversed, and that at the end of the file, a preprocessor warning should be added if the format cannot be built due to missing HAVE_LIBZ.

claudioandre commented:
Since a clean rebase is not possible, I removed the branch. We can revisit this latter.
